File: web/components/dashboard/visualization/format.ts
import type { VisualizationDisplayUnits, VisualizationFormat } from '../../../generated/visualization'
import { decimalMagnitudeOrder, decimalShift, parseDecimal, roundDecimal, type DecimalParts } from './decimal'

export interface ResolvedDisplayUnit {
  scale: number
  suffix: string
  exact: boolean
}

type LocaleData = { decimal: string; group: string; currencySpace: string; currencies: Record<string, string> }
const locales: Record<string, LocaleData> = {
  'en-US': { decimal: '.', group: ',', currencySpace: '', currencies: { USD: '$', BRL: 'R$', EUR: '€' } },
  'pt-BR': { decimal: ',', group: '.', currencySpace: '\u00a0', currencies: { USD: 'US$', BRL: 'R$', EUR: '€' } },
}

export function formatValue(locale: string, format: VisualizationFormat, value: unknown): string {
  const data = locales[locale]
  if (!data) throw new Error(`unsupported visualization locale ${JSON.stringify(locale)}`)
  if (value === null || value === undefined) return '—'
  switch (format.kind) {
    case 'number': return number(data, value, format.minimumFractionDigits ?? 0, format.maximumFractionDigits ?? 3)
    case 'currency': {
      const symbol = data.currencies[format.currency]
      if (!symbol) throw new Error(`unsupported visualization currency ${JSON.stringify(format.currency)}`)
      return symbol + data.currencySpace + number(data, value, format.minimumFractionDigits ?? 2, format.maximumFractionDigits ?? 2)
    }
    case 'percent': return number(data, typeof value === 'string' ? decimalShift(value, 2) : numeric(value) * 100, format.minimumFractionDigits ?? 0, format.maximumFractionDigits ?? 1, '%')
    case 'compact': {
      const numericValue = typeof value === 'string' ? undefined : numeric(value), absolute = typeof value === 'string' ? decimalMagnitude(value) : Math.abs(numericValue!)
      const [scale, suffix] = absolute >= 1e9 ? [1e9, 'B'] : absolute >= 1e6 ? [1e6, 'M'] : absolute >= 1e3 ? [1e3, 'K'] : [1, '']
      return number(data, typeof value === 'string' ? decimalShift(value, -Math.log10(scale)) : numericValue! / scale, 0, format.maximumFractionDigits ?? 1, suffix)
    }
    case 'duration': return duration(data, numeric(value), format.unit)
    case 'temporal': {
      if (typeof value !== 'string' || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z$/.test(value)) throw new Error('temporal visualization value must be RFC 3339 UTC')
      return format.timeStyle && !format.dateStyle ? value.slice(11, 19) : value.slice(0, 10)
    }
  }
}

export function resolveDisplayUnit(policy: VisualizationDisplayUnits, values: readonly unknown[]): ResolvedDisplayUnit {
  if (policy === 'none') return { scale: 1, suffix: '', exact: true }
  const fixed = displayUnit(policy)
  if (fixed) return { ...fixed, exact: false }
  const magnitudes = values.flatMap((value) => {
    if (typeof value === 'number' && Number.isFinite(value)) {
      const absolute = Math.abs(value)
      return absolute === 0 ? [] : [{ order: Math.floor(Math.log10(absolute)), approximate: true, value: absolute }]
    }
    if (typeof value === 'string') {
      const parsed = parseDecimal(value)
      if (!parsed || parsed.integer === '0' && !/[1-9]/.test(parsed.fraction)) return []
      const first = parsed.integer !== '0' ? parsed.integer.length - 1 : -(parsed.fraction.search(/[1-9]/) + 1)
      return [{ order: first, approximate: false, value: 0 }]
    }
    return []
  })
  const maximumOrder = magnitudes.length === 0 ? -Infinity : Math.max(...magnitudes.map((magnitude) => magnitude.order))
  const maximum = magnitudes.filter((magnitude) => magnitude.order === maximumOrder).reduce((max, magnitude) => Math.max(max, magnitude.value), 0)
  const hasExactDecimal = magnitudes.some((magnitude) => !magnitude.approximate)
  let resolved = maximumOrder >= 12 ? displayUnit('trillions')!
    : maximumOrder >= 9 ? displayUnit('billions')!
      : maximumOrder >= 6 ? displayUnit('millions')!
        : maximumOrder >= 3 ? displayUnit('thousands')!
          : { scale: 1, suffix: '' }
  if (!hasExactDecimal && resolved.scale < 1e12 && roundSignificant(maximum / resolved.scale, 3) >= 1000) {
    resolved = resolved.scale === 1 ? displayUnit('thousands')!
      : resolved.scale === 1e3 ? displayUnit('millions')!
        : resolved.scale === 1e6 ? displayUnit('billions')!
          : displayUnit('trillions')!
  }
  return { ...resolved, exact: false }
}

export function resolveDisplayUnitForFormat(
  policy: VisualizationDisplayUnits,
  format: VisualizationFormat | undefined,
  values: readonly unknown[],
): ResolvedDisplayUnit {
  return resolveDisplayUnit(policy, values.map((value) => semanticNumericValue(format, value)))
}

export function formatDisplayValue(
  locale: string,
  format: VisualizationFormat,
  value: unknown,
  unit: ResolvedDisplayUnit,
): string {
  if (unit.exact || format.kind === 'duration' || format.kind === 'temporal') return formatValue(locale, format, value)
  const data = locales[locale]
  if (!data) throw new Error(`unsupported visualization locale ${JSON.stringify(locale)}`)
  if (value === null || value === undefined) return '—'
  const semantic = format.kind === 'percent'
    ? typeof value === 'string' ? decimalShift(value, 2) : numeric(value) * 100
    : value
  if (typeof semantic === 'string') {
    const scaled = decimalShift(semantic, -Math.log10(unit.scale))
    const maximumFractionDigits = decimalSignificantFractionDigits(scaled, 3)
    const suffix = unit.suffix + (format.kind === 'percent' ? '%' : '')
    const formatted = number(data, scaled, 0, maximumFractionDigits, suffix)
    if (format.kind !== 'currency') return formatted
    const symbol = data.currencies[format.currency]
    if (!symbol) throw new Error(`unsupported visualization currency ${JSON.stringify(format.currency)}`)
    return symbol + data.currencySpace + formatted
  }
  const raw = numeric(semantic)
  const scaled = raw / unit.scale
  const maximumFractionDigits = significantFractionDigits(scaled, 3)
  const suffix = unit.suffix + (format.kind === 'percent' ? '%' : '')
  const formatted = number(data, scaled, 0, maximumFractionDigits, suffix)
  if (format.kind !== 'currency') return formatted
  const symbol = data.currencies[format.currency]
  if (!symbol) throw new Error(`unsupported visualization currency ${JSON.stringify(format.currency)}`)
  return symbol + data.currencySpace + formatted
}

function displayUnit(policy: VisualizationDisplayUnits): { scale: number; suffix: string } | undefined {
  switch (policy) {
    case 'thousands': return { scale: 1e3, suffix: 'K' }
    case 'millions': return { scale: 1e6, suffix: 'M' }
    case 'billions': return { scale: 1e9, suffix: 'B' }
    case 'trillions': return { scale: 1e12, suffix: 'T' }
    default: return undefined
  }
}

function semanticNumericValue(format: VisualizationFormat | undefined, value: unknown): unknown {
  if (value === null || value === undefined) return value
  if (typeof value === 'string') return format?.kind === 'percent' ? decimalShift(value, 2) : value
  if (typeof value !== 'number' || !Number.isFinite(value)) return value
  const numericValue = numeric(value)
  return format?.kind === 'percent' ? numericValue * 100 : numericValue
}

function decimalSignificantFractionDigits(value: string, significantDigits: number): number {
  const parsed = parseDecimal(value)
  if (!parsed) throw new Error('visualization value must be a canonical decimal string')
  const first = parsed.integer !== '0' ? parsed.integer.length - 1 : -(parsed.fraction.search(/[1-9]/) + 1)
  return Math.max(0, Math.min(12, significantDigits - first - 1))
}

function significantFractionDigits(value: number, significantDigits: number): number {
  if (value === 0) return 0
  return Math.max(0, Math.min(12, significantDigits - Math.floor(Math.log10(Math.abs(value))) - 1))
}

function roundSignificant(value: number, significantDigits: number): number {
  if (value === 0) return 0
  const factor = 10 ** (significantDigits - Math.floor(Math.log10(Math.abs(value))) - 1)
  return Math.round(value * factor) / factor
}

function number(locale: LocaleData, raw: unknown, minimum: number, maximum: number, suffix = ''): string {
  if (!Number.isInteger(minimum) || !Number.isInteger(maximum) || minimum < 0 || maximum < minimum || maximum > 12) throw new Error(`invalid fraction digit range ${minimum}..${maximum}`)
	if (typeof raw === 'string') return decimalNumber(locale, raw, minimum, maximum, suffix)
	const value = numeric(raw)
  const factor = 10 ** maximum
  const rounded = Math.round(Math.abs(value) * factor) / factor
  let [integer, fraction = ''] = rounded.toFixed(maximum).split('.')
  while (fraction.length > minimum && fraction.endsWith('0')) fraction = fraction.slice(0, -1)
  integer = group(integer, locale.group)
  if (value < 0) integer = `-${integer}`
  return integer + (fraction ? locale.decimal + fraction : '') + suffix
}

function numeric(value: unknown): number {
	if (typeof value === 'number' && Number.isFinite(value)) return value
	throw new Error('visualization value must be a finite number')
}

function decimalNumber(locale: LocaleData, raw: string, minimum: number, maximum: number, suffix: string): string {
  const parsed = parseDecimal(raw)
  if (!parsed) throw new Error('visualization value must be a canonical decimal string')
  const rounded = roundDecimal(parsed, maximum)
  let fraction = rounded.fraction
  while (fraction.length > minimum && fraction.endsWith('0')) fraction = fraction.slice(0, -1)
  const integer = group(rounded.integer, locale.group)
  return (rounded.negative ? '-' : '') + integer + (fraction ? locale.decimal + fraction : '') + suffix
}

function decimalMagnitude(value: string): number {
  const order = decimalMagnitudeOrder(value)
  return order === -Infinity ? 0 : 10 ** order
}

function group(value: string, separator: string): string {
  const first = value.length % 3 || 3
  let output = value.slice(0, first)
  for (let index = first; index < value.length; index += 3) output += separator + value.slice(index, index + 3)
  return output
}

function duration(locale: LocaleData, value: number, unit: string): string {
  if (unit === 'days') return number(locale, value, 0, 1, 'd')
  let seconds = Math.round(value)
  if (unit === 'milliseconds') seconds = Math.round(value / 1000)
  else if (unit === 'minutes') seconds *= 60
  else if (unit === 'hours') seconds *= 3600
  else if (unit !== 'seconds') throw new Error(`unsupported duration unit ${JSON.stringify(unit)}`)
  const hours = Math.floor(seconds / 3600), minutes = Math.floor(seconds % 3600 / 60), secs = seconds % 60
  const parts: string[] = []
  if (hours) parts.push(`${hours}h`); if (minutes) parts.push(`${minutes}m`); if (secs || !parts.length) parts.push(`${secs}s`)
  return parts.join(' ')
}

File: internal/dashboard/visualization/format/format.go
// Package format implements deterministic visualization formatting shared by
// browser-independent renderers and exports.
package format

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/flidai/leapview/internal/dashboard/visualization/ir"
)

type localeData struct{ decimal, group, currencySpace string }

var locales = map[string]localeData{
	"en-US": {decimal: ".", group: ",", currencySpace: ""},
	"pt-BR": {decimal: ",", group: ".", currencySpace: "\u00a0"},
}
var currencies = map[string]map[string]string{
	"en-US": {"USD": "$", "BRL": "R$", "EUR": "€"},
	"pt-BR": {"USD": "US$", "BRL": "R$", "EUR": "€"},
}

func Value(locale string, format ir.VisualizationFormat, value any) (string, error) {
	data, ok := locales[locale]
	if !ok {
		return "", fmt.Errorf("unsupported visualization locale %q", locale)
	}
	if value == nil {
		return "—", nil
	}
	switch spec := format.Value.(type) {
	case *ir.NumberVisualizationFormat:
		return number(data, value, digits(spec.MinimumFractionDigits, 0), digits(spec.MaximumFractionDigits, 3), "")
	case *ir.CurrencyVisualizationFormat:
		symbol, ok := currencies[locale][spec.Currency]
		if !ok {
			return "", fmt.Errorf("unsupported visualization currency %q", spec.Currency)
		}
		formatted, err := number(data, value, digits(spec.MinimumFractionDigits, 2), digits(spec.MaximumFractionDigits, 2), "")
		if err != nil {
			return "", err
		}
		return symbol + data.currencySpace + formatted, nil
	case *ir.PercentVisualizationFormat:
		value, err := numeric(value)
		if err != nil {
			return "", err
		}
		return number(data, value*100, digits(spec.MinimumFractionDigits, 0), digits(spec.MaximumFractionDigits, 1), "%")
	case *ir.CompactVisualizationFormat:
		value, err := numeric(value)
		if err != nil {
			return "", err
		}
		scale, suffix := 1.0, ""
		absolute := math.Abs(value)
		if absolute >= 1_000_000_000 {
			scale, suffix = 1_000_000_000, "B"
		} else if absolute >= 1_000_000 {
			scale, suffix = 1_000_000, "M"
		} else if absolute >= 1_000 {
			scale, suffix = 1_000, "K"
		}
		return number(data, value/scale, 0, digits(spec.MaximumFractionDigits, 1), suffix)
	case *ir.DurationVisualizationFormat:
		value, err := numeric(value)
		if err != nil {
			return "", err
		}
		return duration(data, value, spec.Unit)
	case *ir.TemporalVisualizationFormat:
		text, ok := value.(string)
		if !ok {
			return "", fmt.Errorf("temporal visualization value must be a string")
		}
		if parsed, dateErr := time.Parse(time.DateOnly, text); dateErr == nil {
			if spec.TimeStyle != nil && spec.DateStyle == nil {
				return "", fmt.Errorf("date-only visualization value cannot satisfy a time-only format")
			}
			return parsed.Format(time.DateOnly), nil
		}
		parsed, err := time.Parse(time.RFC3339, text)
		if err != nil {
			return "", fmt.Errorf("parse temporal visualization value: %w", err)
		}
		if spec.TimeStyle != nil && spec.DateStyle == nil {
			return parsed.UTC().Format("15:04:05"), nil
		}
		return parsed.UTC().Format("2006-01-02"), nil
	default:
		return "", fmt.Errorf("unsupported visualization format %T", format.Value)
	}
}

func number(locale localeData, raw any, minimum, maximum int, suffix string) (string, error) {
	value, err := numeric(raw)
	if err != nil {
		return "", err
	}
	if maximum < minimum || minimum < 0 || maximum > 12 {
		return "", fmt.Errorf("invalid fraction digit range %d..%d", minimum, maximum)
	}
	factor := math.Pow10(maximum)
	rounded := math.Round(math.Abs(value)*factor) / factor
	parts := strings.Split(strconv.FormatFloat(rounded, 'f', maximum, 64), ".")
	for len(parts) > 1 && len(parts[1]) > minimum && strings.HasSuffix(parts[1], "0") {
		parts[1] = strings.TrimSuffix(parts[1], "0")
	}
	integer := group(parts[0], locale.group)
	if value < 0 {
		integer = "-" + integer
	}
	if len(parts) == 1 || parts[1] == "" {
		return integer + suffix, nil
	}
	return integer + locale.decimal + parts[1] + suffix, nil
}

func numeric(value any) (float64, error) {
	var number float64
	switch value := value.(type) {
	case float64:
		number = value
	case float32:
		number = float64(value)
	case int:
		number = float64(value)
	case int64:
		number = float64(value)
	case int32:
		number = float64(value)
	default:
		return 0, fmt.Errorf("visualization value %T is not numeric", value)
	}
	if math.IsNaN(number) || math.IsInf(number, 0) {
		return 0, fmt.Errorf("visualization number must be finite")
	}
	return number, nil
}

func group(value, separator string) string {
	if len(value) <= 3 {
		return value
	}
	first := len(value) % 3
	if first == 0 {
		first = 3
	}
	var out strings.Builder
	out.WriteString(value[:first])
	for index := first; index < len(value); index += 3 {
		out.WriteString(separator)
		out.WriteString(value[index : index+3])
	}
	return out.String()
}

func digits(value *int32, fallback int) int {
	if value == nil {
		return fallback
	}
	return int(*value)
}

func duration(locale localeData, value float64, unit string) (string, error) {
	if unit == "days" {
		return number(locale, value, 0, 1, "d")
	}
	seconds := int64(math.Round(value))
	switch unit {
	case "milliseconds":
		seconds = int64(math.Round(value / 1000))
	case "minutes":
		seconds *= 60
	case "hours":
		seconds *= 3600
	case "seconds":
	default:
		return "", fmt.Errorf("unsupported duration unit %q", unit)
	}
	hours, rest := seconds/3600, seconds%3600
	minutes, secs := rest/60, rest%60
	parts := []string{}
	if hours != 0 {
		parts = append(parts, fmt.Sprintf("%dh", hours))
	}
	if minutes != 0 {
		parts = append(parts, fmt.Sprintf("%dm", minutes))
	}
	if secs != 0 || len(parts) == 0 {
		parts = append(parts, fmt.Sprintf("%ds", secs))
	}
	return strings.Join(parts, " "), nil
}
